### DISCUSSION TITLE
Mojo meta-programing: is there an ultimate limitation on how much it can do?

### DISCUSSION THREAD

**Discussion by hcye2022** [modular/modular#3091] [Ideas]
Can Mojo do "full" meta-programming at compile time, e.g., read a file or connect to a server, and include fetched data as constants in the compiled app?

**Comment by Brian-M-J** [comment]
Related to this: if a feature like `embed` is added (see #2977), all this would be possible.

I mean, you wouldn't be opening sockets or reading files at compile time (I'm not sure that's a good idea for security reasons), but you could embed the files themselves in the executable.

You could, for example, have a build script that fetches the file, then compiles the Mojo code with the file embedded in it.
